src/multiplex/client.rs

```diff
@@ -323,6 +323,30 @@ where
     }
 }
 
+/// Handles executing the service error handler in case awaiting the `ClientInner` Future panics.
+struct ClientInnerCleanup<Request, T, E, F>
+where
+    T: Sink<Request> + TryStream,
+    E: From<Error<T, Request>>,
+    F: FnOnce(E),
+{
+    on_service_error: Option<F>,
+    _phantom_data: PhantomData<(Request, T, E)>,
+}
+
+impl<Request, T, E, F> Drop for ClientInnerCleanup<Request, T, E, F>
+where
+    T: Sink<Request> + TryStream,
+    E: From<Error<T, Request>>,
+    F: FnOnce(E),
+{
+    fn drop(&mut self) {
+        if let Some(handler) = self.on_service_error.take() {
+            (handler)(E::from(Error::<T, Request>::TransportUnexpectedExit))
```

Comment:
I wonder if we need to address the case where we are already panicking and the closure panics. Though I can't find the reference code I remember writing for this style of thing. So I think its fine but may be worth some sort of doc comment.

Panic inside a panic causes an abort.

Reply:
Added documentation about the fact that panicking within the closure will abort